### ISSUE TITLE
Android support?

### ISSUE THREAD

**Issue by vzhen** [joshuapinter/react-native-unified-contacts#16]
When ?

**Comment by joshuapinter** [comment]
Working on it right now. Basic support by the end of the weekend, including, checking permissions, reading contacts, etc.

**Comment by vzhen** [comment]
I m ok with basic support as long as with read all contacts.  Big thanks

**Comment by vzhen** [comment]
By the way i hope it support android 4 above.

**Comment by joshuapinter** [comment]
It's currently setup to target a minimum SDK version of 16, which is 4.1. And I'm testing it on a Google Pixel runnin Android 7.

Is the only thing you're going to be using it for is to fetch all Contacts? What else? Selecting a Contact? Etc.? 

Let me know so I can consider it as I'm writing it.

**Comment by joshuapinter** [comment]
@vzhen Basic Android version nearly done. Feel like doing some testing?

**Comment by joshuapinter** [comment]
@vzhen Okay, Android version complete for getting Contacts and checking permissions. Just waiting for you now. Let me know when you're free and I'll give you the instructions.

**Comment by vzhen** [comment]
Sorry for the late reply. 
Yes i just need to fetch all contacts at the moment but may need add, update, delete in future.
I will test and let you know within this week.

**Comment by joshuapinter** [comment]
Cool. I'm just at family dinner now but I'll add some instuctions when I'm back for you.
--
Joshua Pinter
On Sun, Jan 15, 2017 at 5:57 PM, vzhen <notifications@github.com> wrote:
Sorry for the late reply.
Yes i just need to fetch all contacts at the moment but may need add, update, delete in future.
I will test and let you know within this week.

**Comment by joshuapinter** [comment]
@vzhen Couple questions: 

1) Are you searching for contacts by name or just getting ALL the contacts?

2) Do you need access to the Contact's full resolution picture or is their thumbnail image fine?

**Comment by vzhen** [comment]
1. Just need to fetch all the contacts
2. Don't need any picture or thumbnail
But i'm not sure is that possible, i need to listening on new contact added. Which mean if there is a new contact then i will ONLY get the contact.

**Comment by joshuapinter** [comment]
1. No problem. That's ready now. See my next comment for instructions.

2. Okay.

3. Listening for new Contacts is interesting but will have to wait until later. I've never done that before. Maybe after you get it working, you can send a Pull Request to make the library handle it.

**Comment by joshuapinter** [comment]
## Android Instructions

1) Add `"react-native-unified-contacts": "joshuapinter/react-native-unified-contacts#android"` to your `package.json` file and `npm install`.

2) In your console in your project root, run `rnpm link react-native-unified-contacts`. And if that doesn't work, run `react-native link react-native-unified-contacts` instead.

3) Add `import Contacts from 'react-native-unified-contacts';` to your app where you want to use it.

4) Then get all the contacts with this: 

```javascript
Contacts.getContacts()
      .then( function(contacts) { console.log(contacts); } )
      .catch( function(error) { console.log(error); } );
```

Let me know how it goes.

Josh

**Comment by vzhen** [comment]
react-native link didn't include 'react-native-unified-contacts' into `android\setting.gradle`. I add it manually but then another error comes out

`* What went wrong:
A problem occurred configuring project ':react-native-unified-contacts'.
> failed to find target with hash string 'android-25' in: D:\Development\Android\sdk`

**Comment by joshuapinter** [comment]
@vzhen Working on this right now. When will you be able to test it again?

**Comment by joshuapinter** [comment]
@vzhen I know what that error message is saying. You need to install the latest Android SDK, which is API version 25 for Android 7.1.1. Do this through your SDK Manager in Android Studio. Then everything will work. 

Try that and let me know how it goes.

**Comment by geekazoid-at** [comment]
I'd also be very interested in Android support. If I can support you with testing, please let me know!

**Comment by joshuapinter** [comment]
@geekazoid-at You're in luck, it's almost ready. Can you test it out following the instructions here: https://github.com/joshuapinter/react-native-unified-contacts/issues/16#issuecomment-273193105

And let me know how it goes. 

If you get the same error message as @vzhen, it's because you need to update your SDK Manager with Android 7.1.1 for it to compile properly.

Thanks!

**Comment by joshuapinter** [comment]
@geekazoid-at Have you had a chance to test it on Android yet?

**Comment by joonhocho** [comment]
@joshuapinter @geekazoid-at 
I have a fork [joonhocho/react-native-contacts](https://github.com/joonhocho/react-native-contacts), if you want to refer how to get all contact information from Android.
Note that I only use it to get all contact info. Not updates or add.
My suggestion is that if you will be developing for Android, don't try to rename things on native side, but rather try to normalize it on js side.
For example, Android uses `emails` rather than while iOS uses `emailAddresses`.

**Comment by joshuapinter** [comment]
@joonhocho Nice, thanks for posting that. I'll use it to help finish up the Android support thanks.

**Comment by MattFoley** [comment]
@joshuapinter Was any of the Android support ever merged in?

**Comment by joshuapinter** [comment]
@MattFoley Hey Matt. A lot of it has been done but it's sitting on the [android branch](https://github.com/joshuapinter/react-native-unified-contacts/tree/android). 

I've been meaning to finish it up and merge it in but it won't be for a few weeks unless I get help with it. When do you need it by?

**Comment by MattFoley** [comment]
No real hurry, and I don't really have a problem just using your existing branch as is, thanks for the quick response! :)

**Comment by joshuapinter** [comment]
@MattFoley No worries. I'll let you know when it has feature parity. Let me know if you run into any trouble.

**Comment by brettdonald** [comment]
Poke. I am also waiting on Android support.

**Comment by joshuapinter** [comment]
@brettdonald Right on. That's good motivation to get this done. There's currently an `android` branch that's getting close to wrapped up. Any deadline on when you need this, Brett?

**Comment by brettdonald** [comment]
I need something today. And cross-platform support is more important to me than Unified, so I’ll be going with react-native-contacts. Your motivation should be to help the next developer who comes looking for a contacts library.

**Comment by kafein** [comment]
@joshuapinter I'm also waiting. I have about two weeks to finish my project. Is it possible?

Btw, Thanks for your efforts.

**Comment by joshuapinter** [comment]
@kafein You bet. Try installing the current `android` branch and see if you can get the basics working by following the instructions here: https://github.com/joshuapinter/react-native-unified-contacts/issues/16#issuecomment-273193105

So I can focus on the right stuff, what functions are you needing? e.g. Get all contacts. Lookup contact by email, Create contact, Delete contact, etc.

Eventually, we'll get everything in there but just figuring out where I should start so you can start using it in your project.

Thanks.

**Comment by kafein** [comment]
Thanks @joshuapinter 

I'm going to try it. I'll report if i have problems.

**Comment by joshuapinter** [comment]
@Alver23 Check out the `android` branch.

**Comment by Alver23** [comment]
@joshuapinter 
I got this error when executing the command
![image](https://user-images.githubusercontent.com/16018328/36814830-c40dcfd4-1ca6-11e8-8417-36e9b2b5b7d0.png)

**Comment by Alver23** [comment]
![image](https://user-images.githubusercontent.com/16018328/36816023-a2407a24-1caa-11e8-9097-88ba20336e8c.png)

**Comment by joshuapinter** [comment]
@Alver23 Most likely needs to be updated to support the latest version of React Native that remove that method from the super class. If you want to make this update yourself and issue a Pull Request that would be great. 

Otherwise, it'll have to wait until I get around to updating this myself.

**Comment by Alver23** [comment]
@joshuapinter 
I do not have knowledge in java to make the correction. When you could solve it ?

**Comment by joshuapinter** [comment]
@Alver23 Probably not until later in March or maybe April, unfortunately. 

I encourage you to try, though. If you're using React Native, learning the underlying languages of Java and Objective-C/Swift, is extremely helpful.

**Comment by paintedbicycle** [comment]
@joshuapinter What state is this in now? Where do you need it to be to merge it? What’s missing? Can you list out your ideal state for a PR so we can get this moving?

**Comment by joshuapinter** [comment]
@paintedbicycle Ideally we want feature parity (where applicable) with iOS with consistent interfaces between the two. We've been using the Android branch in beta for www.ntwrkapp.com and it's been working great but it's lacking features. Plus, both need to be updated to support the latest versions of iOS/Android/React Native.

**Comment by paintedbicycle** [comment]
@joshuapinter I'm willing to lend a hand, but since it's your repo it would be nice if you look the leadership position. In my opinion since this has been open for over a year, we need to get some momentum behind it by breaking it down into smaller pieces that people can chip away at. Perhaps you can open bug tickets on the Android branch for each thing you'd like to see with a short description? That way we can jump in an help. It would provide us with a roadmap and then we can have discussions on this thread on when it's ok to merge. Most competing libraries do not have full feature parity. I agree it's the goal, but at some point before then merging is likely possible. The opening of tickets and the attention on this branch will help us figure it all out.

**Comment by paintedbicycle** [comment]
Hi everyone, an update on Android support.

I've spent the past week or so trying to align the Android branch with the iOS branch and am starting to get there. If anyone out there can help with the Java, I'd appreciate it as I'm not a Java dev. Basically, we need to take a look at the iOS functions and recreate them all in Java. This has already begun - @joshuapinter got it started by implementing permissions and getContacts.

I've had to make some changes in order to get iOS and Android both working and returning the same data. Basically before the iOS and the Android repos would both work, but if you tried to get them working at the same time with the same JS and in the same ExampeApp, it did not work. So, I have combined the repos on the Android branch (it's now as up-to-date as Master, but also has Android support) and I made some tweaks to get everything up and running.

So:
- I've pull in all changes from the iOS branch of the past year into the Android branch
- I've added Android installation instructions to the Android branch
- I've updated both the iOS and Android apps to ensure that they work (although not all features work in Android yet)
- Note: in order to use the runtime Permissions, we've had to increase the Android SDK version to Android M where those permissions were introduced. (SDK 23). This has been done and since the previous version was already using these functions but without declaring SDK 23 this might have been the source of some of your installation issues. If you follow the example app, both iOS and Android should install perfectly fine.
- Note: if we want to lower the Android dependencies, we can, but I'd need help rewriting the permissions. I'm certainly open to it. Keep in mind that the requirements for iOS are high (iOS 9+) and so perhaps this library is already set up to only support the latest few versions and Android M (Android 6) might fit that perfectly fine. So perhaps we can leave it as it is.

_Android now checks for permissions correctly and returns contacts fine in the same branch as iOS. You can install this branch and start testing as an alpha_

**HELP NEEDED**

High Priority:
- finalize permissions (open settings, request settings in popup)
- writing contacts
- updating contacts
- search string (inside searchContacts function)
- error handling on Android for when an iOS method is called in JS but doesn't exist in Java

Medium Priority:
- deleting contacts

Lower priority (still important):
- groups

In my opinion, Android functions are ok to have breaking changes because Android was never released. We should do everything we can to align Android with the already up and running iOS. Any help is appreciated including testing, PRs or creating unit tests.

Thanks!

**Comment by joshuapinter** [comment]
Excellent work, @paintedbicycle!! Thank you for getting the android branch back up to speed and outlining everything that needs to be done to bring Android to a full parity release with iOS.

I'll be able to help with the high priority items as I'm quite familiar with Java and Android. Now that you have the ExampleApp working for both iOS and Android it should make it easier for other people to jump in and submit a PR as well.

One item I'd like to add to the To-Do list is a decent test suite. I'll take that on to kick off because it's a complicated decision and something that will affect long-term maintenance of the library. This will be critical to confidently accept PRs in the future and for keeping our sanity with our own development.

As per your permissions question, the inspiration of this library was built for the latest contacts framework by Apple that was released with iOS 9. There are other libraries that service the older versions of OS just fine. My vote is keeping this a more modern library. 9+ for iOS and 6+ for Android seem perfectly reasonable.

If others want to add support for older versions, they can submit a PR for us to consider.

My time is limited over the next couple of weeks with business travel and commitments but things should clear up after that. 

However, I'm always available to answer questions and review Pull Requests, etc. 

I'm excited about this moving forward and I want to thank you again @paintedbicycle for all of your hard work thus far. 🙏

**Comment by paintedbicycle** [comment]
Excellent @joshuapinter, thanks. Agreed, testing suite is important.

I've tried quite a few other react-native contacts libraries and while some of them have more complete Android support, there was always something missing from the API (either no postal addresses, or no ability to save contacts, not just read or no searching or the iOS stuff was messy). I think if we can get Android on feature parity with iOS in this repo it will be a very competitive react-native contacts repo and hopefully with that brings more attention and PRs from other devs adding new features.

Except for some smaller tweaks I'm going to step back for a bit now as I've done a big push and need to get back to my other work, but I'm still around and down to help. If we can get the high priority stuff addressed or even one or two more sets of eyes on the Android stuff I think we'll be in good shape.

Paul

**Comment by joshuapinter** [comment]
@paintedbicycle Excellent, thanks again for all your work on this! It'll help push things forward and get it in the spotlight a little more.

**Comment by joshuapinter** [comment]
Thanks for the ping, @paintedbicycle! Yes, let's do this. I can put in a few hours this weekend to get us past this milestone. I'll take a look at your high priority list and start there. 

Thanks again for all your help pushing this along. 👍

**Comment by paintedbicycle** [comment]
Sounds good to me, @joshuapinter. Thanks!

**Comment by joshuapinter** [comment]
I pushed some updates to the Android branch, including updating React and React Native dependencies. I was able to get the Android ExampleApp working with requesting permissions and listing Contacts.

I also added a simple table in the README with a matrix of Platform Compatibility to keep things transparent and organized and to keep us motivated.

Proper Android Support will come in Version 2.0 as it will no doubt require a lot of breaking changes.

One thing I would like to do for Version 2.0 is switch to the Promise structure. Now with `async` and `await` it's a far better programming experience and something we should do. Since this will be a  _major_ breaking change to _all_ the methods, we'll also bundle this into the Version 2.0 release. It shouldn't be too difficult to change. Once we get a good pattern down then it's just a lot of code updating.

In general, I don't like having branches like this open for this length of time so I'm going to merge the `android` branch back into `master` and we can create a 2.0 branch that will include all the breaking changes, including any method normalization, response normalization and Promisification.

This won't affect any current users as we won't update the npm version. The version in `package.json` will be starting at `2.0.0-pre.1` and moving up with pre-releases as we see fit until we are finally able to release `2.0.0` onto npm.

So, I'll go ahead and create that PR now. There's LOTS of work still left to do but I'll try and find bits and pieces of time to chip away at it.
